persistence/docs/SAVEPOINTS_ROLLBACKS.md

```diff
@@ -0,0 +1,146 @@
+- And more...
+
+Performing operations like: copying datadirs, compressing them, etc. is probably not the best approach.
+Having a first-class support for savepoints and rollbacks would be, IMHO, a much better solution.
+
+A more advanced/flexible version of **Savepoints** could be to efficiently serialize the state into an artifact (non-trivial also because we have multiple data-stores but definitely possible), maybe leveraging previous savepoints to only store the changes that have been made since the last one and/or using some version of a WAL (Write-Ahead log that records the changes that happened).
+
+#### Savepoints
+
+A savepoint must have the following properties:
+
+- It must be able to be created at any block height
+- It must be self-contained and/or easy to move around (e.g. a file/archive)
+- The action of creating a savepoint must be atomic (i.e. it must be impossible to create a savepoint that is incomplete)
+- The action of creating a savepoint must be easy to perform (e.g. a CLI command and/or a flag being passed to the node binary when starting it)
+- The action of creating a savepoint must be as fast as possible
+- The operator should be informed with meaningful messages about the progress of the savepoint creation process (telemetry, logging and stdout come to mind)
+- It should be, as much as possible, compact
+- It must have some form of integrity check mechanism (e.g. checksum/hash verification and maybe even a signature that could be very useful in the case of a social rollback)
+
+#### Rollbacks
+
+The following are some of the properties that a rollback mechanism must have:
+
+- If the operator requested a rollback, regardless of the internal state of the node, it must be able to rollback to the requested block height, provided a valid savepoint
+- The rollback process must be atomic (i.e. it must be impossible to rollback to a block height has incomplete/invalid state)
+- The rollback process must be easy to perform (e.g. a CLI command and/or a flag being passed to the node binary when starting it)
+- The rollback process must be as fast as possible
+- The operator should be informed in a meaningful way about the progress of the rollback process (telemetry, logging and stdout come to mind)
+
+
+### Further improvements
+
+- Savepoints could be disseminated/retrieved using other networks like Tor/IPFS/etc., this would free-up bandwidth on the main network and could be used in conjunction with `FastSync` to speed up the process of bootstrapping a new node without overwhelming the `P2P` network with a lot of traffic that is not `Protocol`-related. This could be very important when the network reaches critical mass in terms of scale. Hopefully soon.
+
+  For example: a fresh node could be looking for the latest `Savepoint` signed by PNI available, download it from Tor, apply its state and resume the normal sync process from the other nodes from there.
```

Comment:
`Tor` is great (and the right starting point) but long-term, try to also start thinking in terms of incentive: https://olshansky.medium.com/cryptocurrencies-its-all-about-incentive-77ac47a6adc4

- Is there another protocol that incentives sharing snapshots?
- Could/should we have another protocol actor for this?
- Is it part of Pocket or external?

For example, Gohkan and I were talking about this yesterday: https://olshansky.medium.com/cryptocurrencies-its-all-about-incentive-77ac47a6adc4

We got the idea of having SuperValidators (i.e. higher stake, higher penalty, limited in number) that could help with bootstrapping, and also make the formal 2/3 agreement more flexible. Could be a HotPOKT innovation ;)

Reply:
Yeah, makes sense. Perhaps disseminating snapshots via Tor or other medium is something that I'd start doing as PNI (for example), then I'd look into offloading the responsibility to the protocol itself by creating the necessary incentives but first of all I'd like to see it working like a charm.

Since snapshots/rollback could be seen also as a "disaster recovery" tool as well, I'd advice to use something that's decoupled from Pocket, or at least that can be interacted with in other ways in the case of complete network halt.

I don't know why but I am thinking about Solana now :)

Reply:
Maybe add a "FiremanBoostrapper" or "Node911" as a moonshot idea :)

Reply:
LOL, I'll leave that to you ;)

Reply:
Maybe even an IPFS hash that's stored directly in the block header